tests/azure_iot_hub_client_ut.c

```diff
@@ -381,6 +475,690 @@ static void testAzureIoTHubClient_SendTelemetry_Success( void ** state )
+
+    will_return( AzureIoTMQTT_Subscribe, eAzureIoTMQTTSendFailed );
+    assert_int_equal( AzureIoTHubClient_SubscribeDeviceTwin( &xTestIoTHubClient,
+                                                             prvTestDeviceTwin,
+                                                             NULL, ( uint32_t ) -1 ),
+                      eAzureIoTHubClientSubscribeFailed );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_DeviceTwinSubscribe_ReceiveFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Subscribe, eAzureIoTMQTTSuccess );
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTRecvFailed );
+    assert_int_equal( AzureIoTHubClient_SubscribeDeviceTwin( &xTestIoTHubClient,
+                                                             prvTestDeviceTwin,
+                                                             NULL, ( uint32_t ) -1 ),
+                      eAzureIoTHubClientFailed );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_DeviceTwinSubscribe_Success( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Subscribe, eAzureIoTMQTTSuccess );
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTSuccess );
+    xPacketInfo.ucType = azureiotmqttPACKET_TYPE_SUBACK;
+    xDeserializedInfo.usPacketIdentifier = usTestPacketId;
+    assert_int_equal( AzureIoTHubClient_SubscribeDeviceTwin( &xTestIoTHubClient,
+                                                             prvTestDeviceTwin,
+                                                             NULL, ( uint32_t ) -1 ),
+                      eAzureIoTHubClientSuccess );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_CloudMessageUnsubscribe_InvalidArgFailure( void ** state )
+{
+    ( void ) state;
+
+    /* Fail UnsubscribeCloudToDevice when client is NULL */
+    assert_int_equal( AzureIoTHubClient_UnsubscribeCloudToDeviceMessage( NULL ),
+                      eAzureIoTHubClientInvalidArgument );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_CloudMessageUnsubscribe_SendFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Unsubscribe, eAzureIoTMQTTSendFailed );
+    assert_int_equal( AzureIoTHubClient_UnsubscribeCloudToDeviceMessage( &xTestIoTHubClient ),
+                      eAzureIoTHubClientUnsubscribeFailed );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_CloudMessageUnsubscribe_Success( void ** state )
```

Comment:
I think we should have a test like this, where unsubscribe is called without much setup.  And another test where we go through a successful subscription and then unsubscribe.

I get the way that the state is maintained now it doesn't matter, but this could change in the future and having the test as a backstop like this more faithfully mimics normal app scenario.